### PR TITLE
accel: Add support for udev ACCEL_INITIAL_ORIENTATION property

### PR DESCRIPTION
**Pull request by jwrdegoede**
Hi Bastien,

I know we discussed this a while back and you then said that this should be fixed in the kernel. The problem is that the kernel cannot transparently hide that a lcd is mounted upside-down or 90 degrees rotated. I've been talking to the Intel kernel people about this and hiding this transparently is not possible because not all hardware planes support rotation. E.g. the cursor overlay plane does not support this, so userspace needs to be aware that rotation is being applied and the Xserver or Wayland-compositor needs to pre-rotate the mouse cursor/sprite in software before uploading it to the hardware, which makes it impossible to compensate for a wrongly mounted LCD transparently inside the kernel.

TL;DR: this cannot be fixed in the kernel and needs to be dealt with in userspace, luckily dealing with this is actually quite easy iio-sensor-proxy + the existing rotation code already takes care of everything except for the initial rotation when the device is sitting flat on a table (or other horizontal surface), so this patch adds an udev ACCEL_INITIAL_ORIENTATION property which takes care of this. Note this requires only 6 lines of code :)

As mentioned in the commit message some kernel work is needed to get the text console to not change from the usually proper orientation provided by efifb to the wrong orientation when e.g. the i915 driver loads. There are some patches for this by one of the Intel engineers I will see if I can clean these up one of these days and submit them upstream.

Regards,

Hans

### CODE REVIEW

**Comment by youling257** [comment]
do you know Android x86 project 
		*i7Stylus*)
			set_property hal.sensors.iio.accel.matrix 1,0,0,0,-1,0,0,0,-1
			;;
		*ST70416-6*)
			set_property hal.sensors.iio.accel.matrix 0,-1,0,-1,0,0,0,0,-1
			;;
		*ONDATablet*)
			set_property hal.sensors.iio.accel.matrix 0,1,0,1,0,0,0,0,-1
			;;

**Comment by youling257** [comment]
+	/*
 +	 * DRM_ROTATE_ is counter clockwise to stay compatible with Xrandr
 +	 * while i915 HW rotation is clockwise, thats why this swapping.
 +	 */
 +	switch (val & PLANE_CTL_ROTATE_MASK) {
 +	case PLANE_CTL_ROTATE_0:
 +		break;
 +	case PLANE_CTL_ROTATE_90:
 +		plane_config->rotation = DRM_ROTATE_270;
 +		break;
 +	case PLANE_CTL_ROTATE_180:
 +		plane_config->rotation = DRM_ROTATE_180;
 +		break;
 +	case PLANE_CTL_ROTATE_270:
 +		plane_config->rotation = DRM_ROTATE_90;
 +		break;
 +	}
 +

**Comment by youling257** [comment]
`+	/*
 +	 * DRM_ROTATE_ is counter clockwise to stay compatible with Xrandr
 +	 * while i915 HW rotation is clockwise, thats why this swapping.
 +	 */
 +	switch (val & PLANE_CTL_ROTATE_MASK) {
 +	case PLANE_CTL_ROTATE_0:
 +		break;
 +	case PLANE_CTL_ROTATE_90:
 +		plane_config->rotation = DRM_ROTATE_270;
 +		break;
 +	case PLANE_CTL_ROTATE_180:
 +		plane_config->rotation = DRM_ROTATE_180;
 +		break;
 +	case PLANE_CTL_ROTATE_270:
 +		plane_config->rotation = DRM_ROTATE_90;
 +		break;
 +	}
 +`

**Comment by youling257** [comment]
case DRM_ROTATE_90:
		return PLANE_CTL_ROTATE_270;
	case DRM_ROTATE_180:
		return PLANE_CTL_ROTATE_180;
	case DRM_ROTATE_270:
		return PLANE_CTL_ROTATE_90;
	default:
		MISSING_CASE(rotation);
	}

so needn't 0,1,0,1,0,0,0,0,-1 ？

**Comment by hadess** [comment]
I don't understand what this changes. I would expect the desktop environment not to change the orientation if the screen is already rotated.

Where's the udev patch that adds ACCEL_INITIAL_ORIENTATION ?

**Comment by hadess** [comment]
> I don't understand what this changes. I would expect the desktop environment not to change the orientation if the screen is already rotated.

In fact, this patch will do nothing if the user selected to block the rotation through system menu in GNOME. I don't think that iio-sensor-proxy should come into play at this point.

**Comment by hadess** [comment]
Closing this as the solution should not involve iio-sensor-proxy in any way.
